
import json
import sys
import time
import json
sys.path.append('.')  # NOQA
from location_normalizer import normalize_feature_association
from civic import convert

CIVIC = """
{"civic_actionability_score": null, "evidence_items": [{"status": "submitted", "rating": null, "drug_interaction_type": null, "description": "In a retrospective study of 89 metastatic irinotecan-refractory colorectal cancer patients, patients with a KRAS codon 12 or 13 mutation (n=24) treated with cetuximab plus chemotherapy were associated with reduced response rate (0% vs. 40.0%, P=0.001), shorter progression-free survival (10.1wk, 95% CI:8.0-16.0wk vs. 31.4wk, 95% CI:19.4-36.0wk, P=0.0001, multivariate analysis) and shorter overall survival (10.1mo, 95% CI:5.1-13.0mo vs. 14.3mo, 95% CI:9.4-20.0mo, P=0.026, multivariate analysis) as compared to patients without a KRAS codon 12 or 13 mutation (n=65).", "open_change_count": 0, "evidence_type": "Predictive", "drugs": [{"pubchem_id": null, "id": 16, "name": "Cetuximab"}], "variant_origin": "Somatic Mutation", "disease": {"doid": "9256", "url": "http://www.disease-ontology.org/?id=DOID:9256", "display_name": "Colorectal Cancer", "id": 11, "name": "Colorectal Cancer"}, "source": {"status": "fully curated", "open_access": null, "name": "KRAS mutations as an independent prognostic factor in patients with advanced colorectal cancer treated with cetuximab.", "journal": "J. Clin. Oncol.", "citation": "Li\u00e8vre et al., 2008, J. Clin. Oncol.", "pmc_id": null, "full_journal_title": "Journal of clinical oncology : official journal of the American Society of Clinical Oncology", "source_url": "http://www.ncbi.nlm.nih.gov/pubmed/18202412", "clinical_trials": [], "pubmed_id": "18202412", "is_review": false, "publication_date": {"month": 1, "day": 20, "year": 2008}, "id": 122}, "evidence_direction": "Supports", "variant_id": 148, "clinical_significance": "Resistance or Non-Response", "evidence_level": "C", "type": "evidence", "id": 3703, "name": "EID3703"}], "entrez_name": "KRAS", "variant_types": [{"display_name": "Missense Variant", "description": "A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.", "url": "http://www.sequenceontology.org/browser/current_svn/term/SO:0001583", "so_id": "SO:0001583", "id": 47, "name": "missense_variant"}], "description": "KRAS G12A, like EGFR T790M, has been shown to be capable of driving acquired resistance to tyrosine kinase inhibitors in lung adenocarcinoma.", "entrez_id": 3845, "lifecycle_actions": {"last_commented_on": {"timestamp": "2017-05-15T21:47:42.888Z", "user": {"username": "kkrysiak", "area_of_expertise": "Research Scientist", "twitter_handle": "", "display_name": "kkrysiak", "name": "Kilannin Krysiak", "bio": "Dr. Krysiak is an Instructor at the McDonnell Genome Institute at Washington University School of Medicine where she is involved in the comprehensive genomic analysis of cancer patient cohorts and \u201cn-of-1\u201d studies. She received her PhD in Molecular Genetics and Genomics at Washington University in St. Louis where she focused on the genetics of myelodysplastic syndrome through advanced flow cytometry techniques, primary cell culture and mouse models. She is a founding member of the CIViC team, helping to define the CIViC data model, and a leading content curator and feature development consultant.", "url": "", "created_at": "2015-02-26T04:14:20.953Z", "avatars": {"x32": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=32", "x14": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=14", "x64": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=64", "x128": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=128"}, "orcid": "0000-0002-6299-9230", "accepted_license": null, "affiliation": "", "avatar_url": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=32", "role": "admin", "facebook_profile": "", "linkedin_profile": "kilannin-krysiak-69047819", "organization": {"url": "http://genome.wustl.edu/", "profile_image": {"x32": "/system/organizations/profile_images/000/000/001/x32/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x256": "/system/organizations/profile_images/000/000/001/x256/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x14": "/system/organizations/profile_images/000/000/001/x14/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x64": "/system/organizations/profile_images/000/000/001/x64/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x128": "/system/organizations/profile_images/000/000/001/x128/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976"}, "description": "The McDonnell Genome Institute (MGI) is a world leader in the fast-paced, constantly changing field of genomics. A truly unique institution, we are pushing the limits of academic research by creating, testing, and implementing new approaches to the study of biology with the goal of understanding human health and disease, as well as evolution and the biology of other organisms.", "name": "The McDonnell Genome Institute", "id": 1}, "last_seen_at": "2017-11-28T21:14:34.200Z", "featured_expert": true, "id": 6, "signup_complete": null}}, "last_modified": {"timestamp": "2017-05-15T12:58:41.585Z", "user": {"username": "LynzeyK", "area_of_expertise": "Research Scientist", "twitter_handle": "", "display_name": "LynzeyK", "name": "Lynzey Kujan", "bio": "", "url": "", "created_at": "2015-10-05T13:05:07.668Z", "avatars": {"x32": "https://secure.gravatar.com/avatar/df7a6c4023eebf61d7bd8056368798c0.png?d=identicon&r=pg&s=32", "x14": "https://secure.gravatar.com/avatar/df7a6c4023eebf61d7bd8056368798c0.png?d=identicon&r=pg&s=14", "x64": "https://secure.gravatar.com/avatar/df7a6c4023eebf61d7bd8056368798c0.png?d=identicon&r=pg&s=64", "x128": "https://secure.gravatar.com/avatar/df7a6c4023eebf61d7bd8056368798c0.png?d=identicon&r=pg&s=128"}, "orcid": "orcid.org/0000-0001-9225-7361", "accepted_license": null, "affiliation": "", "avatar_url": "https://secure.gravatar.com/avatar/df7a6c4023eebf61d7bd8056368798c0.png?d=identicon&r=pg&s=32", "role": "curator", "facebook_profile": "", "linkedin_profile": "https://www.linkedin.com/in/lynzey-kujan-135a03a3", "organization": {"url": "http://genome.wustl.edu/", "profile_image": {"x32": "/system/organizations/profile_images/000/000/001/x32/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x256": "/system/organizations/profile_images/000/000/001/x256/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x14": "/system/organizations/profile_images/000/000/001/x14/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x64": "/system/organizations/profile_images/000/000/001/x64/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x128": "/system/organizations/profile_images/000/000/001/x128/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976"}, "description": "The McDonnell Genome Institute (MGI) is a world leader in the fast-paced, constantly changing field of genomics. A truly unique institution, we are pushing the limits of academic research by creating, testing, and implementing new approaches to the study of biology with the goal of understanding human health and disease, as well as evolution and the biology of other organisms.", "name": "The McDonnell Genome Institute", "id": 1}, "last_seen_at": "2017-11-22T21:12:17.431Z", "featured_expert": false, "id": 83, "signup_complete": null}}, "last_reviewed": {"timestamp": "2017-05-15T21:47:42.888Z", "user": {"username": "kkrysiak", "area_of_expertise": "Research Scientist", "twitter_handle": "", "display_name": "kkrysiak", "name": "Kilannin Krysiak", "bio": "Dr. Krysiak is an Instructor at the McDonnell Genome Institute at Washington University School of Medicine where she is involved in the comprehensive genomic analysis of cancer patient cohorts and \u201cn-of-1\u201d studies. She received her PhD in Molecular Genetics and Genomics at Washington University in St. Louis where she focused on the genetics of myelodysplastic syndrome through advanced flow cytometry techniques, primary cell culture and mouse models. She is a founding member of the CIViC team, helping to define the CIViC data model, and a leading content curator and feature development consultant.", "url": "", "created_at": "2015-02-26T04:14:20.953Z", "avatars": {"x32": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=32", "x14": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=14", "x64": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=64", "x128": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=128"}, "orcid": "0000-0002-6299-9230", "accepted_license": null, "affiliation": "", "avatar_url": "https://secure.gravatar.com/avatar/17180f9afc9f7f04fff97197c1ee5cb6.png?d=identicon&r=pg&s=32", "role": "admin", "facebook_profile": "", "linkedin_profile": "kilannin-krysiak-69047819", "organization": {"url": "http://genome.wustl.edu/", "profile_image": {"x32": "/system/organizations/profile_images/000/000/001/x32/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x256": "/system/organizations/profile_images/000/000/001/x256/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x14": "/system/organizations/profile_images/000/000/001/x14/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x64": "/system/organizations/profile_images/000/000/001/x64/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976", "x128": "/system/organizations/profile_images/000/000/001/x128/MGI_STANDARD4_logo_brown-example_v1b.png?1494525976"}, "description": "The McDonnell Genome Institute (MGI) is a world leader in the fast-paced, constantly changing field of genomics. A truly unique institution, we are pushing the limits of academic research by creating, testing, and implementing new approaches to the study of biology with the goal of understanding human health and disease, as well as evolution and the biology of other organisms.", "name": "The McDonnell Genome Institute", "id": 1}, "last_seen_at": "2017-11-28T21:14:34.200Z", "featured_expert": true, "id": 6, "signup_complete": null}}}, "assertions": [], "provisional_values": {}, "gene_id": 30, "name": "G12A", "variant_groups": [{"variants": [{"entrez_id": 1956, "variant_types": [{"display_name": "Missense Variant", "description": "A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.", "url": "http://www.sequenceontology.org/browser/current_svn/term/SO:0001583", "so_id": "SO:0001583", "id": 47, "name": "missense_variant"}], "description": "EGFR T790M was one of the very first mutations recognized to confer resistance to targeted therapies in non-small cell lung cancer. While successful in amplified EGFR, the efficacy of the first and second generation TKI's (erlotinib, gefitinib, neratinib) in treating patients harboring this mutation before treatment is notably lower. This lack of efficacy can likely be to blame for the poorer prognosis for patients with this mutation as compared to patients with wildtype EGFR or other types of EGFR mutations. Approximately half of EGFR mutant tumors with acquired resistance to TKI inhibition have been shown to harbor this mutation, implicating it as a mechanism of acquired therapy resistence. A third generation TKI (osimertinib) has been approved for the treatment of EGFR T790M mutant NSCLC. Patients positive for T790M in a plasma-based test have similar outcomes like those with tumor biopsy testing.", "gene_id": 19, "name": "T790M", "entrez_name": "EGFR", "coordinates": {"chromosome2": null, "reference_bases": "C", "start2": null, "variant_bases": "T", "stop": 55249071, "stop2": null, "representative_transcript2": null, "start": 55249071, "representative_transcript": "ENST00000275493.2", "ensembl_version": 75, "chromosome": "7", "reference_build": "GRCh37"}, "type": "variant", "id": 34, "civic_actionability_score": null}, {"entrez_id": 3845, "variant_types": [{"display_name": "Missense Variant", "description": "A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.", "url": "http://www.sequenceontology.org/browser/current_svn/term/SO:0001583", "so_id": "SO:0001583", "id": 47, "name": "missense_variant"}], "description": "While the KRAS G12 region is a widely studied recurrent region in cancer, its impact on clinical action is still debated. Often associated with tumors that are wild-type for other drivers (EGFR and ALK specifically), the prognosis for patients with this mutation seems to be worse than the KRAS wild-type cohort in patients with colorectal and pancreatic cancer, however this hypothesis is in need of further validation. This mutation, along with the mutations affecting the neighboring G13 position, may result in a less responsive tumor when treated with first-generation TKI's like gefitinib. However, cetuximab treatment was shown to extend survival in a cohort of colorectal patients.", "gene_id": 30, "name": "G12C", "entrez_name": "KRAS", "coordinates": {"chromosome2": null, "reference_bases": "C", "start2": null, "variant_bases": "A", "stop": 25398285, "stop2": null, "representative_transcript2": null, "start": 25398285, "representative_transcript": "ENST00000256078.4", "ensembl_version": 75, "chromosome": "12", "reference_build": "GRCh37"}, "type": "variant", "id": 78, "civic_actionability_score": null}, {"entrez_id": 3845, "variant_types": [{"display_name": "Missense Variant", "description": "A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.", "url": "http://www.sequenceontology.org/browser/current_svn/term/SO:0001583", "so_id": "SO:0001583", "id": 47, "name": "missense_variant"}], "description": "While the KRAS G12 region is a widely studied recurrent region in cancer, its impact on clinical action is still actively debated. Often associated with tumors that are wild-type for other drivers (EGFR and ALK specifically), the prognosis for patients with this mutation seems to be worse than the KRAS wild-type cohort in patients with colorectal and pancreatic cancer, however this hypothesis is in need of further validation. This mutation, along with the mutations affecting the neighboring G13 position, may result in a less responsive tumor when treated with first-generation TKI's like gefitinib. The NCCN guidelines for colorectal cancer contain recommendations that the targeted therapies cetuximab and panitumumab should only be used in the context of wild type KRAS. However, cetuximab treatment was shown to extend survival in a single cohort of colorectal patients with G12D mutations. Overall, the interpretation for KRAS mutations in most clinical scenarios is still undecided.", "gene_id": 30, "name": "G12D", "entrez_name": "KRAS", "coordinates": {"chromosome2": null, "reference_bases": "C", "start2": null, "variant_bases": "T", "stop": 25398284, "stop2": null, "representative_transcript2": null, "start": 25398284, "representative_transcript": "ENST00000256078.4", "ensembl_version": 75, "chromosome": "12", "reference_build": "GRCh37"}, "type": "variant", "id": 79, "civic_actionability_score": null}, {"entrez_id": 3845, "variant_types": [{"display_name": "Missense Variant", "description": "A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.", "url": "http://www.sequenceontology.org/browser/current_svn/term/SO:0001583", "so_id": "SO:0001583", "id": 47, "name": "missense_variant"}], "description": "KRAS G12A, like EGFR T790M, has been shown to be capable of driving acquired resistance to tyrosine kinase inhibitors in lung adenocarcinoma.", "gene_id": 30, "name": "G12A", "entrez_name": "KRAS", "coordinates": {"chromosome2": null, "reference_bases": "C", "start2": null, "variant_bases": "G", "stop": 25398284, "stop2": null, "representative_transcript2": null, "start": 25398284, "representative_transcript": "ENST00000256078.4", "ensembl_version": 75, "chromosome": "12", "reference_build": "GRCh37"}, "type": "variant", "id": 148, "civic_actionability_score": null}], "type": "variant_group", "description": "EGFR pathway activation is a nearly ubiquitous hallmark of cancer. Many tyrosine kinase inhibitors have been developed to target EGFR pathway activity. One such inhibitor, erlotinib, has demonstrated efficacy in an EGFR over-active setting. However, the T790M missense mutation has shown to confer resistance to this inhibitor in cell lines and case studies.  ", "name": "EGFR TKI Resistance", "id": 12}], "sources": [], "allele_registry_id": "CA135567", "variant_aliases": ["RS121913529", "GLY12ALA"], "hgvs_expressions": ["NM_004985.4:c.35G>C", "NP_004976.2:p.Gly12Ala", "NC_000012.11:g.25398284C>G", "ENST00000256078.4:c.35G>C"], "errors": {}, "coordinates": {"chromosome2": null, "reference_bases": "C", "start2": null, "variant_bases": "G", "stop": 25398284, "stop2": null, "representative_transcript2": null, "start": 25398284, "representative_transcript": "ENST00000256078.4", "ensembl_version": 75, "chromosome": "12", "reference_build": "GRCh37"}, "type": "variant", "id": 148, "clinvar_entries": ["45122"]}
"""

CIVIC2 = """
{"civic_actionability_score": null, "evidence_items": [{"status": "accepted", "rating": 3, "drug_interaction_type": null, "description": "A patient with metastatic colorectal cancer was enrolled in a phase 2 clinical trial (NCT01174121). 24 cultures of tumor-infiltrating lymphocytes were expanded from 3 lung metastasis resections. After cultures were tested for reactivity, the patient was infused with 1.48 x 10^11 tumor-infiltrating lymphocytes (75% CD8+ T cells) that were specifically reactive to KRAS G12D and HLA-C*08:02. A single lesion that progressed after 9 months still expressed the KRAS G12D mutation. Whole-exome analysis revealed copy-neutral loss of heterozygosity at chromosome 6, encoding the HLA locus and resulting in loss of the targeted HLA-C*08:02 allele.", "open_change_count": 0, "evidence_type": "Predictive", "drugs": [{"pubchem_id": null, "id": 432, "name": "Adoptive T-cell Transfer"}], "variant_origin": "Somatic Mutation", "disease": {"doid": "9256", "url": "http://www.disease-ontology.org/?id=DOID:9256", "display_name": "Colorectal Cancer", "id": 11, "name": "Colorectal Cancer"}, "source": {"status": "fully curated", "open_access": true, "name": "T-Cell Transfer Therapy Targeting Mutant KRAS in Cancer.", "journal": "N. Engl. J. Med.", "citation": "Tran et al., 2016, N. Engl. J. Med.", "pmc_id": "PMC5178827", "full_journal_title": "The New England journal of medicine", "source_url": "http://www.ncbi.nlm.nih.gov/pubmed/27959684", "clinical_trials": [], "pubmed_id": "27959684", "is_review": false, "publication_date": {"month": 12, "day": 8, "year": 2016}, "id": 1330}, "evidence_direction": "Supports", "variant_id": 821, "clinical_significance": "Resistance or Non-Response", "evidence_level": "C", "type": "evidence", "id": 1899, "name": "EID1899"}], "entrez_name": "HLA-C", "variant_types": [], "description": "", "entrez_id": 3107, "lifecycle_actions": {}, "assertions": [], "provisional_values": {}, "gene_id": 2608, "name": "COPY-NEUTRAL LOSS OF HETEROZYGOSITY", "variant_groups": [], "sources": [], "allele_registry_id": null, "variant_aliases": [], "hgvs_expressions": [], "errors": {}, "coordinates": {"chromosome2": null, "reference_bases": null, "start2": null, "variant_bases": null, "stop": null, "stop2": null, "representative_transcript2": null, "start": null, "representative_transcript": null, "ensembl_version": null, "chromosome": null, "reference_build": null}, "type": "variant", "id": 821, "clinvar_entries": []}
"""


def test_civic():
    evidence = json.loads(CIVIC)
    gene_data = {'gene': 'KRAS', 'civic': {'variants': [evidence]}}

    fa = convert(gene_data).next()
    normalize_feature_association(fa)
    for f in fa['features']:
        print f['name'], f['provenance']
    assert len(fa['features']) == 1


def test_civic2():
    evidence = json.loads(CIVIC2)
    gene_data = {'gene': 'KRAS', 'civic': {'variants': [evidence]}}

    fa = convert(gene_data).next()
    normalize_feature_association(fa)
    for f in fa['features']:
        print f['name'], f['provenance']
    assert len(fa['features']) == 1
    assert fa['features'][0]['geneSymbol'] == "HLA-C"
    assert fa['features'][0]['provenance_rule'] == "is_loss"
